projects/igniteui-angular/src/lib/services/transaction/utilities.ts

```diff
@@ -0,0 +1,88 @@
+export enum TransactionType {
+    ADD = 'add',
+    DELETE = 'delete',
+    UPDATE = 'update'
+}
+
+export interface ITransaction {
+    id: any;
+    type: TransactionType;
+    newValue: any;
+}
+
+export interface IState {
+    value: any;
+    recordRef: any;
+    type: TransactionType;
+}
+
+export interface IgxTransactionService {
+    /**
+     * Adds provided  transaction with recordRef if any
+     * @param transaction Transaction to be added
+     * @param recordRef Reference to the value of the record in the data source related to the changed item
+     *
+     * @returns If operation is successful.
+     */
+    add(transaction: ITransaction, recordRef?: any): boolean;
+
+    /**
+     * Returns an array of all transactions. If id is provided returns last transaction for provided id
+     */
+    getTransactionLog(id?: any): ITransaction[] | ITransaction;
+
+    /**
+     * Remove the last transaction if any
+     */
+    undo(): void;
+
+    /**
+     * Applies the last undone transaction if any
+     */
+    redo(): void;
+
+    /**
+     * Returns a map of aggregated state for all transactions
+     */
+    aggregatedState(): Map<any, IState>;
```

Comment:
I have a feeling this would be much better served for the API as `ITransaction[]` return.
The Map doesn't serialize out of the box and the `Array.from(map)` result is a jagged array which is no good.
That means the `aggregatedState()` will have to do some conversions so the end result is easily usable and can be sent to a server.

PS: No need to describe the type of result ("Returns a map"), there's a type right there. The test should describe what 'aggregated' means - all changes are aggregated resulting in a single Transaction per record that has a value reflecting the final state of the row.

Reply:
All this is fixed now